rest/indextest/index_test.go

```diff
@@ -596,6 +596,226 @@ func TestSyncOnline(t *testing.T) {
+	resp.RequireStatus(http.StatusOK)
+
+	resp = rest.BootstrapAdminRequest(t, sc, http.MethodGet, "/"+keyspace+"/_config/sync", "")
+	resp.RequireResponse(http.StatusOK, syncFunc)
+
+	// Attempt to delete sync function while in starting mode
+	resp = rest.BootstrapAdminRequest(t, sc, http.MethodDelete, "/"+keyspace+"/_config/sync", "")
+	resp.RequireStatus(http.StatusOK)
+
+	resp = rest.BootstrapAdminRequest(t, sc, http.MethodGet, "/"+keyspace+"/_config/sync", "")
+	resp.RequireResponse(http.StatusOK, "")
+
+	// Take the database back online while async init is still in progress, verify state goes to Starting
+	resp = rest.BootstrapAdminRequest(t, sc, http.MethodPost, "/"+dbName+"/_config", string(dbOnlineConfigPayload))
+	resp.RequireStatus(http.StatusCreated)
+	waitAndRequireDBState(t, sc, dbName, db.DBStarting)
+
+	// Unblock initialization, verify status goes to Online
+	close(unblockInit)
+	waitAndRequireDBState(t, sc, dbName, db.DBOnline)
+
+}
+
+// TestAsyncInitRemoteConfigUpdates verifies that a database with in-progress async
+// index initialization can accept config updates made on other nodes (arriving via polling)
+// (prior to CBG-4008, operations would block waiting for async init to complete)
+func TestAsyncInitRemoteConfigUpdates(t *testing.T) {
+	if base.UnitTestUrlIsWalrus() {
+		t.Skip("This test only works against Couchbase Server")
+	}
+	base.TestRequiresCollections(t)
+	base.SetUpTestLogging(t, base.LevelDebug, base.KeyHTTP)
+
+	// enable config polling to allow testing of cross-node updates
+	bootstrapConfig := rest.BootstrapStartupConfigForTest(t)
+	bootstrapConfig.Bootstrap.ConfigUpdateFrequency = base.NewConfigDuration(1 * time.Second)
+	sc, closeFn := rest.StartServerWithConfig(t, &bootstrapConfig)
+	defer closeFn()
+
+	// Set testing callbacks for async initialization
+	collectionCount := int64(0)
+	initStarted := make(chan error)
+	unblockInit := make(chan error)
+	collectionCompleteCallback := func(dbName, collectionName string) {
+		count := atomic.AddInt64(&collectionCount, 1)
+		// On first collection, close initStarted channel
+		log.Printf("collection callback count: %v", count)
+		if count == 1 {
+			log.Printf("closing initStarted")
+			close(initStarted)
+		}
+		rest.WaitForChannel(t, unblockInit, "waiting for test to unblock initialization")
+	}
+	sc.DatabaseInitManager.SetCallbacks(collectionCompleteCallback, nil)
+
+	ctx := base.TestCtx(t)
+	// Get a test bucket, and use it to create the database.
+	tb := base.GetTestBucket(t)
+	defer tb.Close(ctx)
+
+	importFilter := "function(doc) { return true }"
+	syncFunc := "function(doc){ channel(doc.channels); }"
+
+	dbName := "db"
+	dbConfig := makeDbConfig(t, tb, syncFunc, importFilter)
+	dbConfig.Name = dbName
+	dbConfig.StartOffline = base.BoolPtr(true)
+
+	keyspace := dbName
+	if len(dbConfig.Scopes) > 0 {
+		keyspaces := getRESTKeyspaces(dbName, dbConfig.Scopes)
+		keyspace = keyspaces[0]
+	}
```

Comment:
```suggestion
		keyspace := getRESTKeyspaces(dbName, dbConfig.Scopes)[0]
```

This is always > 0, in fact it is always 3.

Reply:
Do you mean this is always 1 when base.TestRequiresCollections(t) passes?

Reply:
I think it is be three because `dbConfig := makeDbConfig(t, tb, syncFunc, importFilter)` creates 3 collections always.